### PR TITLE
Fix a Focus.onKey bug and update some docs

### LINKED ISSUE #83023: FocusNode onKey function is not called
## Steps to Reproduce

<details>
  <summary>Minimal code sample</summary>

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Shortcuts(
      shortcuts: <LogicalKeySet, Intent>{
        LogicalKeySet(LogicalKeyboardKey.select): const ActivateIntent(),
      },
      child: MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Row(
          children: [
            MainImageBox(key: ValueKey('asdf')),
          ],
        ),
      ),
    );
  }
}

class MainImageBox extends StatefulWidget {
  const MainImageBox({required Key key}) : super(key: key);

  @override
  _MainImageBoxState createState() => _MainImageBoxState();
}

class _MainImageBoxState extends State<MainImageBox> {
  late FocusNode focusNode;

  @override
  void initState() {
    super.initState();
    focusNode = FocusNode(onKey: handleKey);
  }

  bool handleKey(FocusNode focusNode, RawKeyEvent key) {
    print("handleKey function called");
    return false;
  }

  @override
  Widget build(BuildContext context) {
    return RawMaterialButton(
      focusNode: focusNode,
      focusColor: Colors.transparent,
      splashColor: Colors.transparent,
      onPressed: () async {
        print('press');
      },
      child: Container(
        color: Colors.green,
        width: 230,
        padding: EdgeInsets.fromLTRB(5.0, 0.0, 5.0, 0.0),
      ),
    );
  }

  @override
  void dispose() {
    focusNode.dispose();
    super.dispose();
  }
}

```

</details>

1. Run `Minimal code sample`.
2. Move focus around with arrow keys

**Expected results:** 
FocusNode onKey function should be called

**Actual results:** 
onKey function is not called

Works on Flutter 2.0.2 but not 2.2

<details>
  <summary>Logs</summary>

```
➜  onkey flutter analyze
Analyzing onkey...                                                      
No issues found! (ran in 2.9s)
➜  onkey 

```

```
➜  onkey flutter doctor -v
[✓] Flutter (Channel stable, 2.2.0, on macOS 11.3.1 20E241 darwin-x64, locale en-GB)
    • Flutter version 2.2.0 at /Users/snaebjorn/development/flutter
    • Framework revision b22742018b (6 days ago), 2021-05-14 19:12:57 -0700
    • Engine revision a9d88a4d18
    • Dart version 2.13.0

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/snaebjorn/Library/Android/sdk
    • Platform android-S, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2018.2.4)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 29.0.4
    • Dart plugin version 182.4323.44

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • sdk google atv x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • macOS (desktop)             • macos         • darwin-x64     • macOS 11.3.1 20E241 darwin-x64
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 90.0.4430.212
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code -13)
    ! Error: Living Room is not connected. Xcode will continue when Living Room is connected. (code -13)
    ! Error: Auglysingadeild4thgen is not connected. Xcode will continue when Auglysingadeild4thgen is connected. (code -13)
    ! Error: Snaeji is not connected. Xcode will continue when Snaeji is connected. (code -13)

! Doctor found issues in 1 category.
➜  onkey 

```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes https://github.com/flutter/flutter/issues/83023

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -1027,7 +1028,8 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
   @mustCallSuper
   FocusAttachment attach(BuildContext? context, {FocusOnKeyCallback? onKey}) {
     _context = context;
-    this.onKey = onKey ?? this.onKey;
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/focus_manager.dart:1030]: I also thought about it for a long time before I realized the problem, haha

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 2]
```diff
@@ -645,7 +645,7 @@ class _FocusState extends State<Focus> {
     }());
 
     if (oldWidget.focusNode == widget.focusNode) {
-      if (widget.onKey != focusNode.onKey) {
+      if (widget.onKey != null && widget.onKey != focusNode.onKey) {
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/focus_scope.dart:648]:
```suggestion
      if ((widget.onKey != null && widget.onKey != focusNode.onKey) ||
           (widget.onKey == null && oldWidget.onKey != null)) {
```

File: packages/flutter/test/widgets/focus_scope_test.dart [hunk 3]
```diff
@@ -1625,6 +1625,54 @@ void main() {
...
+
+      keyEventHandled = null;
+      await tester.pumpWidget(
+        Focus(
+          onKey: (_, __) => KeyEventResult.ignored, // This one does nothing.
+          focusNode: focusNode,
+          child: Container(key: key1),
+        ),
+      );
+      expect(keyEventHandled, null);
```
Comment (xu-baolin) [inline packages/flutter/test/widgets/focus_scope_test.dart:1674]:
```suggestion
      expect(keyEventHandled, null);
      
            await tester.pumpWidget(
        Focus(
          onKey: null, // Remove this.
          focusNode: focusNode,
          child: Container(key: key1),
        ),
      );
      expect(keyEventHandled, true);
```

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1030]: @gspencergoog Hi, I wanna ask what should we do if both `Focus.onKey` and `FocusNode.onKey` are non-null?

**Review by gspencergoog — COMMENTED** [review]
I actually don't think we need to make this change, I think we need to make a documentation change.

Instead of setting the `onKey` handler on a focus node, in most cases a new `Focus` widget should be introduced and the `onKey` set on that instead. The callbacks on a `FocusNode` should really only be set if you're not using `Focus` widgets (e.g. in a render object, or managing it yourself in a `StatefulWidget`). I realize this is not well documented.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1030]: Event the `Focus.onKey` is null?

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 4]
```diff
@@ -645,7 +645,8 @@ class _FocusState extends State<Focus> {
     }());
 
     if (oldWidget.focusNode == widget.focusNode) {
-      if (widget.onKey != focusNode.onKey) {
+      if ((widget.onKey != null && widget.onKey != focusNode.onKey) ||
+          (widget.onKey == null && oldWidget.onKey != null)) {
```
[the remark this answers is not in the document]
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_scope.dart:649]: You are right if the `Focus.onKey` has a higher priority even it is null.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_manager.dart:1030]: @gspencergoog Hi,
The processing of the first build and subsequent update processing is inconsistent. 
The former is to rewrite `node.onKey` when `Focus.onKey` is not null, but the latter is unconditional rewriting. 
This test case can reproduce this problem.
```dart
    testWidgets("xxx", (WidgetTester tester) async {
      final GlobalKey key1 = GlobalKey(debugLabel: '1');
      bool? keyEventHandled;
      final FocusNode focusNode = FocusNode(
        onKey: (FocusNode node, RawKeyEvent event) {
          keyEventHandled = true;
          return KeyEventResult.handled;
        }
      );

      await tester.pumpWidget(
        Focus(
          onKey: null,
          focusNode: focusNode,
          child: Container(key: key1),
        ),
      );

      Focus.of(key1.currentContext!).requestFocus();
      await tester.pump();
      await tester.sendKeyEvent(LogicalKeyboardKey.enter);
      expect(keyEventHandled, true);

      keyEventHandled = null;
      // Update the Focus widget.
      await tester.pumpWidget(
        Focus(
          onKey: null,
          focusNode: focusNode,
          child: Container(key: key1),
        ),
      );

      await tester.sendKeyEvent(LogicalKeyboardKey.enter);
      expect(keyEventHandled, true);
    });
```

**Comment by xu-baolin** [comment]
First build processing, rewrite the `node.onKey` when `focus.onKey` is non-null:
```dart
  FocusAttachment attach(BuildContext? context, {FocusOnKeyCallback? onKey}) {
    _context = context;
    this.onKey = onKey ?? this.onKey;
    _attachment = FocusAttachment._(this);
    return _attachment!;
  }
```
Updating processing, unconditional rewriting with the `focus.onKey`, even it is null:
```dart
    if (oldWidget.focusNode == widget.focusNode) {
      if (widget.onKey != focusNode.onKey) {
        focusNode.onKey = widget.onKey;
      }
```
Finally, we need to confirm when `Focus.onKey` is null, do we need to rewrite the `node.onKey`? CC @gspencergoog

**Comment by gspencergoog** [comment]
Yeah, @xu-baolin you are right that it is inconsistent. It is probably better to overwrite the `onKey` at initialization with whatever is in the `Focus` widget, since the `Focus` widget is supposed to be managing the focus node, and the current initialization behavior is inconsistent with that. That will be a breaking change, though, since there are probably people relying on the current behavior.

And we do need some appropriate documentation around how it should be used. I'm working on that documentation.

File: packages/flutter/test/widgets/focus_manager_test.dart [hunk 5]
```diff
@@ -895,9 +895,9 @@ void main() {
       final FocusScopeNode scope2 = FocusScopeNode(debugLabel: 'Scope 2');
       final FocusAttachment scope2Attachment = scope2.attach(context, onKey: handleEvent);
       final FocusNode parent1 = FocusNode(debugLabel: 'Parent 1', onKey: handleEvent);
-      final FocusAttachment parent1Attachment = parent1.attach(context);
+      final FocusAttachment parent1Attachment = parent1.attach(context, onKey: parent1.onKey);
       final FocusNode parent2 = FocusNode(debugLabel: 'Parent 2', onKey: handleEvent);
-      final FocusAttachment parent2Attachment = parent2.attach(context);
+      final FocusAttachment parent2Attachment = parent2.attach(context, onKey: parent2.onKey);
```
Comment (xu-baolin) [inline packages/flutter/test/widgets/focus_manager_test.dart:900]: You are right, this change actually breaks some tests.

**Comment by xu-baolin** [comment]
CC @gspencergoog

**Comment by Piinks** [comment]
(triage) @gspencergoog can you take a look at this PR?

**Comment by gspencergoog** [comment]
I'm still working on the documentation article for this (an article about how to use the Focus system).

This is likely to be a breaking change that is pretty subtle: it won't cause any compile errors, some things will just stop working the way they used to.

Can we throw an assert or an exception when someone overwrites the `onKey` with the value in the widget when they didn't mean to (i.e. when they are relying on the old behavior)? I'm worried that it might give a lot of false positives, however.

**Comment by xu-baolin** [comment]
> I'm still working on the documentation article for this (an article about how to use the Focus system).
> 
> This is likely to be a breaking change that is pretty subtle: it won't cause any compile errors, some things will just stop working the way they used to.
> 
> Can we throw an assert or an exception when someone overwrites the `onKey` with the value in the widget when they didn't mean to (i.e. when they are relying on the old behavior)? I'm worried that it might give a lot of false positives, however.

I will try it!

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 6]
```diff
@@ -575,6 +575,11 @@ class _FocusState extends State<Focus> {
     _canRequestFocus = focusNode.canRequestFocus;
     _descendantsAreFocusable = focusNode.descendantsAreFocusable;
     _hasPrimaryFocus = focusNode.hasPrimaryFocus;
+
+    assert(
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/focus_scope.dart:579]: @gspencergoog Hey, please review, thanks.

**Comment by xu-baolin** [comment]
> I'm still working on the documentation article for this (an article about how to use the Focus system).
> 
> This is likely to be a breaking change that is pretty subtle: it won't cause any compile errors, some things will just stop working the way they used to.
> 
> Can we throw an assert or an exception when someone overwrites the `onKey` with the value in the widget when they didn't mean to (i.e. when they are relying on the old behavior)? I'm worried that it might give a lot of false positives, however.

Done.

**Comment by skia-gold** [comment]
Gold has detected about 16 new digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/83233

**Comment by skia-gold** [comment]
Gold has detected about 16 new digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/83233

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/83233

**Comment by Piinks** [comment]
Hey @xu-baolin, there were some flaky golden file tests earlier this week, but they have been reverted. Updating here should resolve those warnings. :)

**Comment by xu-baolin** [comment]
Continue working via #87564

**Pull request closed without merging**